shm CHANGE allow setting a custom prefix for SHM files

rjarry opened the pull request:
Allow exporting a `SYSREPO_SHM_PREFIX` environment variable. When that variable is set, all files created under `/dev/shm` will use that prefix. If unset, the files wille have their "default" names.

This can be useful for testing purposes, or to run multiple "groups" of sysrepo applications in concordance with the `SYSREPO_REPOSITORY_PATH` environment variable.

michalvasko commented:
Hi Robin,
I get the idea and I think it should be possible to implement but in a different way. You see, SHM files are created with `shm_open(3)`, which will always ignore whatever prefix you want to set, it always uses the same one (and it should be `/dev/shm`). The only way to allow several *sysrepo* instances would be to add a dir after the base path so you get `/dev/shm/dir/sr_main`, for example. At least I think it should be possible to use a prefix in the name given to `shm_open()`.

Regards,
Michal

rjarry commented:
Hi Michal,

unfortunately, this is not possible. `shm_open(3)` (on linux) states:

> The operation of `shm_open()` is analogous to that of `open(2)`.  `name` specifies the shared memory object  to  be created or opened.  For portable use, a shared memory object should be identified by a name of the form `/somename`; that is, a null-terminated string of up to `NAME_MAX` (i.e., 255) characters consisting  of  an  initial slash, followed by one or more characters, __none of which are slashes__.

The POSIX man page is also conservative about that:

> It is unspecified  whether the  `name`  appears in the file system and is visible to other functions that take pathnames as arguments. The `name` argument conforms to the construction rules for a pathname, except that the  interpretation  of  <slash> characters  other  than  the leading <slash> character in name is implementation-defined, and that the length limits for the name argument are implementation-defined and need not be  the  same  as  the  pathname  limits {PATH_MAX} and {NAME_MAX}.  If name begins with the <slash> character, then processes calling `shm_open()` with the same value of name refer to the same shared memory object, as long as that name has not been removed.  If name does not begin with the <slash> character, the effect is implementation-defined.

The idea behind this patch is to use for example:

```bash
export SYSREPO_SHM_PREFIX="group1-"
```

And it will create `/dev/shm/group1-*` files.

michalvasko commented:
I see, okay, then what I suggested is not possible what in the pull request you have not changed names for `shm_open()` at all so it would not work the way you said it should. Also, all *sysrepo* SHM files use either `sr_` or `srsub_` prefix (unless I forgot something), which are currently hard-coded so the best solution would probably be to unify all these prefixes to be returned by a function that can get them from the environment.

Regards,
Michal

rjarry commented:
Unless I made a mistake, the `sr_` or `srsub_` prefixes are still here. `SYSREPO_SHM_PREFIX` (if defined) is only prepended before these.

Here is an example:

```bash
~# SYSREPO_SHM_PREFIX="group1-" ./application &
~# ls -l /dev/shm/group1-*
-rw-rw-rw- 1 root root 1.9K 07-01 13:42 /dev/shm/group1-sr_ext
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-datastores.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-datastores.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-netconf-notifications.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-netconf-notifications.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-netconf.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-netconf.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-netconf-with-defaults.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-netconf-with-defaults.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-origin.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-origin.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-yang-library.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_ietf-yang-library.running
-rw-rw-rw- 1 root root 6.5K 07-01 13:42 /dev/shm/group1-sr_main
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_sysrepo-monitoring.operational
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_sysrepo-monitoring.running
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_mymodule.operational
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_mymodule.running
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1-srsub_mymodule.oper.c2a945a3
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_yang.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1-sr_yang.running
```

Do you mean you would like to use the same `sr` prefix by default for all SHM files and change it if `SYSREPO_SHM_PREFIX` is set?

michalvasko commented:
Oh, sorry, my bad, the prefix is always used. Okay, so please put getting the prefix into separate function like I said before and use it on these 2 places. It should probably also be checked in the function that the prefix does not include any `/`.

Regards,
Michal

rjarry commented:
Ok I'll do that.

What do you think about completely replacing the `sr` prefix when `SYSREPO_SHM_PREFIX` is set? That would give something like this:

With `SYSREPO_SHM_PREFIX="group1"`:

```
-rw-rw-rw- 1 root root 1.9K 07-01 13:42 /dev/shm/group1_ext
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-datastores.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-datastores.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-netconf-notifications.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-netconf-notifications.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-netconf.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-netconf.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-netconf-with-defaults.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-netconf-with-defaults.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-origin.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-origin.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-yang-library.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_ietf-yang-library.running
-rw-rw-rw- 1 root root 6.5K 07-01 13:42 /dev/shm/group1_main
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1_sysrepo-monitoring.operational
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1_sysrepo-monitoring.running
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1_mymodule.operational
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1_mymodule.running
-rw------- 1 root root    7 07-01 13:42 /dev/shm/group1_sub_mymodule.oper.c2a945a3
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_yang.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/group1_yang.running
```

And if SYSREPO_SHM_PREFIX is not set:

```
-rw-rw-rw- 1 root root 1.9K 07-01 13:42 /dev/shm/sr_ext
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-datastores.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-datastores.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-netconf-notifications.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-netconf-notifications.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-netconf.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-netconf.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-netconf-with-defaults.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-netconf-with-defaults.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-origin.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-origin.running
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-yang-library.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_ietf-yang-library.running
-rw-rw-rw- 1 root root 6.5K 07-01 13:42 /dev/shm/sr_main
-rw------- 1 root root    7 07-01 13:42 /dev/shm/sr_sysrepo-monitoring.operational
-rw------- 1 root root    7 07-01 13:42 /dev/shm/sr_sysrepo-monitoring.running
-rw------- 1 root root    7 07-01 13:42 /dev/shm/sr_mymodule.operational
-rw------- 1 root root    7 07-01 13:42 /dev/shm/sr_mymodule.running
-rw------- 1 root root    7 07-01 13:42 /dev/shm/sr_sub_mymodule.oper.c2a945a3
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_yang.operational
-rw-rw-rw- 1 root root    7 07-01 13:42 /dev/shm/sr_yang.running
```

Note that this would change `srsub_` to `sr_sub_` for subscriptions.

rjarry commented:
Or I can skip the additional `_` to preserve existing file names.

michalvasko commented:
I guess it is fine to completely replace the prefix but I would prefer to keep it the way it is if the prefix is not set. You then always have `sr_main` and `sr_ext`, which are custom shared memory chunks, `sr_*`, which are LYB files, and `srsub_*`, which are subscription shared memory chunks. Changing it to `sr_sub_*` could cause collisions with modules starting with `sub_`.

Regards,
Michal

rjarry commented:
It took me more time than I expected. I did not understand what you were saying:

> which are currently hard-coded

It should be OK now.

michalvasko commented:
Hi Robin,
I am sorry but I do not like the current changes. I am not even sure what exactly your use-case is when you need to have several independent *sysrepo* instances but the added complexity is not worth it in my opinion. To learn from the code the final path for some SHM file was difficult enough, now it is almost impossible. So, unless you can somehow simplify it (even I can attempt it once I have some time), I would prefer to keep it the way it is now without this feature.

Regards,
Michal

rjarry commented:
Hi Michal,

The main use-case is parallel testing and automated builds. The big issue here is that you cannot (per POSIX standards) control where the SHM files are created by `shm_open()`. If you have a build system (for example, [buildroot](https://buildroot.org/)) that needs to call `sysrepoctl` and/or `sysrepocfg` commands to install YANG modules and prepare startup configs into a staging directory (before generation of a `bzImage`), without my change there is no way to have multiple builds in parallel. Not to mention the issue if the build host is already running sysrepo applications.

This change allows some kind of isolation by using unique `SYSREPO_SHM_PREFIX` values for each separate build or test run. For example I am planning to use this feature for unit tests of the sysrepo python bindings I am working on.

I tried hard to find another solution to trick `shm_open()` into using a specific path but that is not possible without **major** `LD_PRELOAD` hacks. If you give me guidance on how to implement this in a simpler way, I can adapt my code.

Thanks

michalvasko commented:
Okay, fair enough. So these are my ideas that may make the code better.

There would be 2 prefixes, one `sr_` by default, the other `srsub_` by default. There would be a function to retrieve each of them. In your case you could set both prefixes to be the same but the distinction should be kept.

As for `main` and `ext` SHM, there could be again 2 functions that would generate their path and they would use the `sr_` prefix. I think that may make it a bit clearer in the code what is actually happenning (`sr_path_global_shm` seems too vague to me).

Also, please follow the rest of the code conventions in always returning `sr_error_info_t *`, NULL on success and the check can be together with the function call for these short prototypes.

Regards,
Michal

rjarry commented:
Ok I can try that however, having 2 separate prefixes means we need to export two variables. It seems a bit tedious to me. Why not use a single variable and append `sub` for subscriptions?

michalvasko commented:
Okay, then the format can be `PREFIX(sub)_NAME` meaning with the default prefix `sr` you get `sr_NAME` and `srsub_NAME` (just like it is now). Meaning the underscore will be hard-coded and not part of the prefix.

rjarry commented:
Hi Michal, is that ok like this?

michalvasko commented on src/common.c:
```diff
@@ -1418,16 +1418,81 @@ sr_create_module_imps_incs_r(const struct lys_module *ly_mod)
+        *path = NULL;
+    }
+
+end:
```
Please do not use redundant labels, this one is not needed at all.

rjarry replied:
I removed it.

michalvasko commented on src/common.c:
```diff
@@ -1418,16 +1418,81 @@ sr_create_module_imps_incs_r(const struct lys_module *ly_mod)
+        *path = NULL;
+    }
+
+end:
```
Another redundant label.

rjarry replied:
I removed it.

michalvasko commented on src/common.c:
```diff
@@ -1418,16 +1418,81 @@ sr_create_module_imps_incs_r(const struct lys_module *ly_mod)
+
     if (suffix2 > -1) {
-        ret = asprintf(path, "%s/srsub_%s.%s.%08x", abs_path ? SR_SHM_DIR : "", mod_name, suffix1, (uint32_t)suffix2);
+        ret = asprintf(path, "%s/%ssub_%s.%s.%08x",
```
The convention I used is that in case a line is too long (more than 120 chars), I continue on the next line starting with double indentation (8 spaces). There are several similar places where this is different.

rjarry replied:
done

michalvasko commented on src/sysrepo.c:
```diff
@@ -352,8 +352,17 @@ sr_connect(const sr_conn_options_t opts, sr_conn_ctx_t **conn_p)
-            shm_unlink(SR_MAIN_SHM);
-            shm_unlink(SR_EXT_SHM);
+            char *shm_name = NULL;
+            if (sr_path_main_shm(&shm_name) == NULL) {
```
Please use `!` instead of `== NULL` just like in the rest of the code.

rjarry replied:
done

rjarry commented:
@michalvasko I have no issue in enforcing coding style consistency, don't worry :)

I think I have addressed all your comments. Please tell me if I need to change something else.

Maybe we could set up some sort of automatic coding style checks (with [clang-format](https://clang.llvm.org/docs/ClangFormat.html) for example), what do you think?

michalvasko commented on src/sysrepo.c:
```diff
@@ -352,8 +352,17 @@ sr_connect(const sr_conn_options_t opts, sr_conn_ctx_t **conn_p)
-            shm_unlink(SR_MAIN_SHM);
-            shm_unlink(SR_EXT_SHM);
+            char *shm_name = NULL;
+            if (!sr_path_main_shm(&shm_name)) {
```
Actually, I did not realize that this is wrong because in the unlikely event that this fails, the `err_info` will leak. In cases when the error should not be propagated, you should just free it, it is printed when it is generated.

rjarry replied:
right, I'll fix that

rjarry replied:
Can I use sr_errinfo_merge ?

michalvasko replied:
Yes, I only looked just now that it is executed only when some error occurred so I guess it makes more sense to also return it.

rjarry replied:
I should be OK now.

michalvasko commented on src/sysrepo.c:
```diff
@@ -352,8 +352,17 @@ sr_connect(const sr_conn_options_t opts, sr_conn_ctx_t **conn_p)
+                free(shm_name);
+                shm_name = NULL;
+            }
+            if (!sr_path_ext_shm(&shm_name)) {
```
Similar leak.

michalvasko commented:
Well, we have some automatic code formatting in our wishlist for quite some time now but not having a high priority, it is consistently postponed. But if you can help us with it, we would appreciate it.

Regards,
Michal

rjarry commented:
> Well, we have some automatic code formatting in our wishlist for quite some time now but not having a high priority, it is consistently postponed. But if you can help us with it, we would appreciate it.

I'll see if I find some time.

michalvasko commented on src/common.c:
```diff
@@ -1418,16 +1418,79 @@ sr_create_module_imps_incs_r(const struct lys_module *ly_mod)
+    } else if (strchr(*prefix, '/') != NULL) {
+        *prefix = NULL;
+        sr_errinfo_new(&err_info, SR_ERR_INVAL_ARG, NULL,
+            SR_SHM_PREFIX_ENV" cannot contain slashes.");
```
One last formatting problem.

rjarry replied:
done

michalvasko commented on src/sysrepo.c:
```diff
@@ -352,8 +352,22 @@ sr_connect(const sr_conn_options_t opts, sr_conn_ctx_t **conn_p)
+            } else {
+                shm_unlink(shm_name);
+                free(shm_name);
+                shm_name = NULL;
```
Seems that this line is not needed.

rjarry replied:
What line exactly? If `sr_path_main_shm` failed, `shm_name` is `NULL`.

michalvasko replied:
But it does not matter whether `shm_name` is NULL or not in the code. I would just like to keep this code short because it is only error handling.

rjarry replied:
I don't understand what I need to change here.

If `shm_name` is `NULL`, `shm_unlink` should not be called. If it is not `NULL`, it must be freed after calling `shm_unlink`. Did you have something else in mind?

michalvasko replied:
We seem not to understand each other, all I am saying is that the code `shm_name = NULL;` seems redundant to me and can be removed. If you want to keep it consistent then I would make sure the functions either allocate `shm_name` on success or set it to `NULL` on error. Maybe even `shm_name` initialization can be removed but since it is not taking up another line, I leave that up to you.

rjarry replied:
We did not understand each other, indeed :) I'll remove that.

michalvasko commented on src/sysrepo.c:
```diff
@@ -352,8 +352,22 @@ sr_connect(const sr_conn_options_t opts, sr_conn_ctx_t **conn_p)
+            } else {
+                shm_unlink(shm_name);
+                free(shm_name);
+                shm_name = NULL;
```
Same as above.

rjarry commented:
The checks on ppc64le are failing because of some lock timeout. I'm not sure it is related to my changes here. However could it be related to #2033?

michalvasko commented:
Well, the timeouts are because of the unhandled locks, yes. But the problem is caused by some other, not intentional, lock problem and it is not related to your changes. Not sure if I can fix it, though, because it seems to depend on some data race as well.

Regards,
Michal

rjarry commented:
Thanks Michal.
